src/dns/query.cpp

```diff
@@ -3,46 +3,306 @@
+#include <measurement_kit/common/logger.hpp>
 #include <measurement_kit/common/poller.hpp>
+#include <measurement_kit/common/settings.hpp>
+#include "src/common/utils.hpp"
 #include <measurement_kit/common/var.hpp>
 
+#include <event2/dns.h>
+
+#include <cassert>
 #include <functional>
 #include <iosfwd>
+#include <map>
+#include <new>
+#include <stdexcept>
 #include <string>
 #include <type_traits>
 
-#include "src/dns/query_impl.hpp"
+#include <netinet/in.h>
+#include <sys/socket.h>
+#include <arpa/inet.h>
 
 struct evdns_base;
 
 namespace mk {
+namespace dns {
 
-class Logger;
 
-namespace dns {
+class QueryContext {
+    public:
+        ~QueryContext() {
+            if (base != nullptr) {
+                evdns_base_free(base, 0);
+            }
+        }
+
+        double ticks;
+
+        evdns_base *base = nullptr;
+
+        Message message;
+        std::function<void(Error, Message)> callback;
+
+        QueryContext(evdns_base *b,
+                std::function<void(Error, Message)> c,
+                     Message m) {
+            base = b;
+            callback = c;
+            message = m;
+            ticks = mk::time_now();
+        }
+};
+
+static evdns_base *create_evdns_base(Settings settings,
+        Poller *poller = mk::get_global_poller()) {
+
+    evdns_base *base;
+    event_base *evb = poller->get_event_base();
+
+    // XXX free variables when entering error path
+    if (settings.find("nameserver") != settings.end()) {
+        if ((base = evdns_base_new(evb, 0)) == nullptr) {
+            throw std::bad_alloc();
+        }
+        if (evdns_base_nameserver_ip_add(
+                base, settings["nameserver"].c_str()) != 0) {
+            throw std::runtime_error("Cannot set server address");
+        }
+    } else if ((base = evdns_base_new(evb, 1)) == nullptr) {
+        throw std::bad_alloc();
+    }
+
+    if (settings.find("attempts") != settings.end() &&
+        evdns_base_set_option(base, "attempts",
+                                    settings["attempts"].c_str()) != 0) {
+        throw std::runtime_error("Cannot set 'attempts' option");
+    }
+    if (settings.find("timeout") != settings.end() &&
+        evdns_base_set_option(base, "timeout",
+                                    settings["timeout"].c_str()) != 0) {
+        throw std::runtime_error("Cannot set 'timeout' option");
+    }
 
-Query::Query(QueryClass dns_class, QueryType dns_type, std::string name,
-             std::function<void(Error, Response)> func, Logger *lp,
-             evdns_base *dnsb, Libs *libs) {
-    if (dnsb == nullptr) {
-        dnsb = mk::get_global_evdns_base();
+    // By default we don't randomize the query's case
+    std::string randomiz{"0"};
+    if (settings.find("randomize_case") != settings.end()) {
+        randomiz = settings["randomize_case"];
     }
-    if (libs == nullptr) {
-        libs = Libs::global();
+    if (evdns_base_set_option(base, "randomize-case", randomiz.c_str()) !=
+        0) {
```

Comment:
Also here you should `evdns_base_free(base, 1);` to free the allocated `evdns_base`.